README.md
# ncpbst
Dies ist eine Testversion von NCP, um ein docker-iamge mit btrfs support zu
erstellen

nur bei Erfolg soll eine Vorschlag an die original Distribution erfolgen




changelog.md

[v1.20.3](https://github.com/BstAA/nextcloudpi/commit/1378192) (2020-01-04) repo in nextcloudpi umbenannt


install.sh
#!/bin/bash

# NextCloudPi installation script
#
# Copyleft 2017 by Ignacio Nunez Hernanz <nacho _a_t_ ownyourbits _d_o_t_ com>
# GPL licensed (see end of file) * Use at your own risk!
#
# Usage: ./install.sh
#
# more details at https://ownyourbits.com

BRANCH=master
#DBG=x

set -e$DBG

TMPDIR="$(mktemp -d /tmp/nextcloudpi.XXXXXX || (echo "Failed to create temp dir. Exiting" >&2 ; exit 1) )"
trap "rm -rf \"${TMPDIR}\" ; exit 0" 0 1 2 3 15

[[ ${EUID} -ne 0 ]] && {
  printf "Must be run as root. Try 'sudo $0'\n"
  exit 1
}

export PATH="/usr/local/sbin:/usr/sbin:/sbin:${PATH}"

# check installed software
type mysqld  &>/dev/null && echo ">>> WARNING: existing mysqld configuration will be changed <<<"

# get install code
echo "Getting build code..."
apt-get update
apt-get install --no-install-recommends -y wget ca-certificates sudo lsb-release

pushd "$TMPDIR"
wget -qO- --content-disposition https://github.com/BstAA/nextcloudpi/archive/"$BRANCH"/latest.tar.gz \
  | tar -xz \
  || exit 1
cd - && cd "$TMPDIR"/nextcloudpi-"$BRANCH"

# install NCP
echo -e "\nInstalling NextCloudPi..."
source etc/library.sh

# check distro
check_distro etc/ncp.cfg || {
  echo "ERROR: distro not supported:";
  cat /etc/issue
  exit 1;
}


mkdir -p /usr/local/etc/ncp-config.d/
cp etc/ncp-config.d/nc-nextcloud.cfg /usr/local/etc/ncp-config.d/
cp etc/library.sh /usr/local/etc/
cp etc/ncp.cfg /usr/local/etc/

install_app    lamp.sh
install_app    bin/ncp/CONFIG/nc-nextcloud.sh
run_app_unsafe bin/ncp/CONFIG/nc-nextcloud.sh
systemctl restart mysqld # TODO this shouldn't be necessary, but somehow it's needed in Debian 9.6. Fixme
install_app    ncp.sh
run_app_unsafe bin/ncp/CONFIG/nc-init.sh
bash /usr/local/bin/ncp-provisioning.sh
bash /usr/local/bin/ncp-update

popd

IFACE="$( ip r | grep "default via" | awk '{ print $5 }' | head -1 )"
IP="$( ip a show dev "$IFACE" | grep global | grep -oP '\d{1,3}(.\d{1,3}){3}' | head -1 )"

echo "Done.

First: Visit https://$IP/  https://nextcloudpi.local/ (also https://nextcloudpi.lan/ or https://nextcloudpi/ on windows and mac)
to activate your instance of NC, and save the auto generated passwords. You may review or reset them
anytime by using nc-admin and nc-passwd.
Second: Type 'sudo ncp-config' to further configure NCP, or access ncp-web on https://$IP:4443/
Note: You will have to add an exception, to bypass your browser warning when you
first load the activation and :4443 pages. You can run letsencrypt to get rid of
the warning if you have a (sub)domain available.
"

exit 0

# License
#
# This script is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This script is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this script; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330,
# Boston, MA  02111-1307  USA


buildlib.sh
#!/bin/bash

# Library to install software on Raspbian ARM through QEMU
#
# Copyleft 2017 by Ignacio Nunez Hernanz <nacho _a_t_ ownyourbits _d_o_t_ com>
# GPL licensed (see end of file) * Use at your own risk!
#
# More at ownyourbits.com
#

DBG=x

# $IMG    is the source image
# $IP     is the IP of the QEMU images
# $IMGOUT will contain the name of the generated image
function launch_install_qemu()
{
  local IMG=$1
  local IP=$2
  [[ "$IP"      == ""  ]] && { echo "usage: launch_install_qemu <img> <IP>"; return 1; }
  test -f "$IMG"          || { echo "input file $IMG not found";             return 1; }

  IMGOUT="$IMG-$( date +%s )"
  cp --reflink=auto -v "$IMG" "$IMGOUT" || return 1

  pgrep qemu-system-arm &>/dev/null && { echo -e "QEMU instance already running. Abort..."; return 1; }
  launch_qemu "$IMGOUT" &
  sleep 10
  wait_SSH "$IP"
  launch_installation_qemu "$IP" || return 1 # uses $INSTALLATION_CODE
  wait 
  echo "$IMGOUT generated successfully"
}

function launch_qemu()
{
  local IMG=$1
  test -f "$1" || { echo "Image $IMG not found"; return 1; }
  test -d qemu-raspbian-network || git clone https://github.com/nachoparker/qemu-raspbian-network.git
  sed -i '30s/NO_NETWORK=1/NO_NETWORK=0/' qemu-raspbian-network/qemu-pi.sh
  sed -i '35s/NO_GRAPHIC=0/NO_GRAPHIC=1/' qemu-raspbian-network/qemu-pi.sh
  echo "Starting QEMU image $IMG"
  ( cd qemu-raspbian-network && sudo ./qemu-pi.sh ../"$IMG" 2>/dev/null )
}

function ssh_pi()
{
  local IP=$1
  local ARGS=${@:2}
  local PIUSER=${PIUSER:-pi}
  local PIPASS=${PIPASS:-raspberry}
  local SSH=( ssh -q  -o UserKnownHostsFile=/dev/null\
                      -o StrictHostKeyChecking=no\
                      -o ServerAliveInterval=20\
                      -o ConnectTimeout=20\
                      -o LogLevel=quiet                  )
  type sshpass &>/dev/null && local SSHPASS=( sshpass -p$PIPASS )
  if [[ "${SSHPASS[@]}" == "" ]]; then
    ${SSH[@]} ${PIUSER}@$IP $ARGS; 
  else
    ${SSHPASS[@]} ${SSH[@]} ${PIUSER}@$IP $ARGS 
    local RET=$?
    [[ $RET -eq 5 ]] && { ${SSH[@]} ${PIUSER}@$IP $ARGS; return $?; }
    return $RET
  fi
}

function wait_SSH()
{
  local IP=$1
  echo "Waiting for SSH to be up on $IP..."
  while true; do
    ssh_pi "$IP" : && break
    sleep 1
  done
  echo "SSH is up"
}

function launch_installation()
{
  local IP=$1
  [[ "$INSTALLATION_CODE"  == "" ]] && { echo "Need to run config first"    ; return 1; }
  [[ "$INSTALLATION_STEPS" == "" ]] && { echo "No installation instructions"; return 1; }
  local PREINST_CODE="
set -e$DBG
sudo su
set -e$DBG
"
  echo "Launching installation"
  echo -e "$PREINST_CODE\n$INSTALLATION_CODE\n$INSTALLATION_STEPS" | ssh_pi "$IP" || { echo "Installation to $IP failed" && return 1; }
}

function launch_installation_qemu()
{
  local IP=$1
  [[ "$NO_CFG_STEP"  != "1" ]] && local CFG_STEP=configure
  [[ "$NO_CLEANUP"   != "1" ]] && local CLEANUP_STEP="if [[ \$( type -t cleanup ) == function ]];then cleanup; fi"
  [[ "$NO_HALT_STEP" != "1" ]] && local HALT_STEP="nohup halt &>/dev/null &"
  local INSTALLATION_STEPS="
install
$CFG_STEP
$CLEANUP_STEP
$HALT_STEP
"
  launch_installation "$IP" # uses $INSTALLATION_CODE
}

function launch_installation_online()
{
  local IP=$1
  [[ "$NO_CFG_STEP" != "1" ]]  && local CFG_STEP=configure
  local INSTALLATION_STEPS="
install
$CFG_STEP
"
  launch_installation "$IP" # uses $INSTALLATION_CODE
}

function prepare_dirs()
{
  [[ "$CLEAN" == "0" ]] || rm -rf cache
  rm -rf tmp
  mkdir -p tmp output cache
}

function mount_raspbian()
{
  local IMG="$1"
  local MP=raspbian_root

  [[ -f "$IMG"        ]] || { echo "no image";  return 1; }
  [[ -e "$MP" ]] && { echo "$MP already exists"; return 1; }

  local SECTOR=$( fdisk -l "$IMG" | grep Linux | awk '{ print $2 }' )
  local OFFSET=$(( SECTOR * 512 ))
  mkdir -p "$MP"
  sudo mount $IMG -o offset=$OFFSET "$MP" || return 1
  echo "Raspbian image mounted"
}

function mount_raspbian_boot()
{
  local IMG="$1"
  local MP=raspbian_boot

  [[ -f "$IMG" ]] || { echo "no image";           return 1; }
  [[ -e "$MP"  ]] && { echo "$MP already exists"; return 1; }

  local SECTOR=$( fdisk -l "$IMG" | grep FAT32 | awk '{ print $2 }' )
  local OFFSET=$(( SECTOR * 512 ))
  mkdir -p "$MP"
  sudo mount $IMG -o offset=$OFFSET "$MP" || return 1
  echo "Raspbian image mounted"
}

function umount_raspbian()
{
  [[ -d raspbian_root ]] || [[ -d raspbian_boot ]] || { echo "Nothing to umount"; return 0; }
  [[ -d raspbian_root ]] && { sudo umount -l raspbian_root; rmdir raspbian_root || return 1; }
  [[ -d raspbian_boot ]] && { sudo umount -l raspbian_boot; rmdir raspbian_boot || return 1; }
  echo "Raspbian image umounted"
}

function prepare_chroot_raspbian()
{
  local IMG="$1"
  mount_raspbian "$IMG" || return 1
  sudo mount -t proc proc     raspbian_root/proc/
  sudo mount -t sysfs sys     raspbian_root/sys/
  sudo mount -o bind /dev     raspbian_root/dev/
  sudo mount -o bind /dev/pts raspbian_root/dev/pts

  sudo cp /usr/bin/qemu-arm-static raspbian_root/usr/bin

  # Prevent services from auto-starting
  sudo bash -c "echo -e '#!/bin/sh\nexit 101' > raspbian_root/usr/sbin/policy-rc.d"
  sudo chmod +x raspbian_root/usr/sbin/policy-rc.d
}

function clean_chroot_raspbian()
{
  sudo rm -f raspbian_root/usr/bin/qemu-arm-static
  sudo rm -f raspbian_root/usr/sbin/policy-rc.d
  sudo umount -l raspbian_root/{proc,sys,dev/pts,dev}
  umount_raspbian
}

# sets DEV
function resize_image()
{
  local IMG="$1"
  local SIZE="$2"
  local DEV
  echo -e "\n\e[1m[ Resize Image ]\e[0m"
  fallocate -l$SIZE "$IMG"
  parted "$IMG" -- resizepart 2 -1s
  DEV="$( sudo losetup -f )"
  mount_raspbian "$IMG"
  sudo resize2fs -f "$DEV"
  echo "Image resized"
  umount_raspbian
}

function update_boot_uuid()
{
  local IMG="$1"
  local PTUUID="$( sudo blkid -o export "$IMG" | grep PTUUID | sed 's|.*=||' )"

  echo -e "\n\e[1m[ Update Raspbian Boot UUIDS ]\e[0m"
  mount_raspbian "$IMG" || return 1
  sudo bash -c "cat > raspbian_root/etc/fstab" <<EOF
PARTUUID=${PTUUID}-01  /boot           vfat    defaults          0       2
PARTUUID=${PTUUID}-02  /               ext4    defaults,noatime  0       1
EOF
  umount_raspbian

  mount_raspbian_boot "$IMG"
  sudo bash -c "sed -i 's|root=[^[:space:]]*|root=PARTUUID=${PTUUID}-02 |' raspbian_boot/cmdline.txt"
  umount_raspbian
}

function prepare_sshd()
{
  local IMG="$1"
  mount_raspbian_boot "$IMG" || return 1
  sudo touch raspbian_boot/ssh   # this enables ssh
  umount_raspbian
}

function set_static_IP()
{
  local IMG="$1"
  local IP="$2"
  mount_raspbian "$IMG" || return 1
  sudo bash -c "cat > raspbian_root/etc/dhcpcd.conf" <<EOF
interface eth0
static ip_address=$IP/24
static routers=192.168.0.1
static domain_name_servers=8.8.8.8

# Local loopback
auto lo
iface lo inet loopback
EOF
  umount_raspbian
}

function copy_to_image()
{
  local IMG=$1
  local DST=$2
  local SRC=${@: 3 }

  mount_raspbian "$IMG" || return 1
  sudo cp --reflink=auto -v "$SRC" raspbian_root/"$DST" || return 1
  sync
  umount_raspbian
}

function deactivate_unattended_upgrades()
{
  local IMG=$1

  mount_raspbian "$IMG" || return 1
  sudo rm -f raspbian_root/etc/apt/apt.conf.d/20ncp-upgrades
  umount_raspbian
}

function download_raspbian()
{
  local IMGFILE=$1
  local IMG_CACHE=cache/raspbian_lite.img
  local ZIP_CACHE=cache/raspbian_lite.zip

  echo -e "\n\e[1m[ Download Raspbian ]\e[0m"
  mkdir -p cache
  test -f $IMG_CACHE && \
    echo -e "INFO: $IMG_CACHE already exists. Skipping download ..." && \
    cp -v --reflink=auto $IMG_CACHE "$IMGFILE" && \
    return 0

  test -f "$ZIP_CACHE" && {
    echo -e "INFO: $ZIP_CACHE already exists. Skipping download ..."
  } || {
    wget https://downloads.raspberrypi.org/raspbian_lite_latest -O "$ZIP_CACHE" || return 1
  }

  unzip -o "$ZIP_CACHE" && \
    mv *-raspbian-*.img $IMG_CACHE && \
    cp -v --reflink=auto $IMG_CACHE "$IMGFILE" 
}

function pack_image()
{
  local IMG="$1"
  local TAR="$2"
  local DIR="$( dirname  "$IMG" )"
  local IMGNAME="$( basename "$IMG" )"
  echo -e "\n\e[1m[ Pack Image ]\e[0m"
  echo "packing $IMG → $TAR"
  tar -I pbzip2 -C "$DIR" -cvf "$TAR" "$IMGNAME" && \
    echo -e "$TAR packed successfully"
}

function create_torrent()
{
  local TAR="$1"
  echo -e "\n\e[1m[ Create Torrent ]\e[0m"
  [[ -f "$TAR" ]] || { echo "image $TAR not found"; return 1; }
  local IMGNAME="$( basename "$TAR" .tar.bz2 )"
  local DIR="torrent/$IMGNAME"
  [[ -d "$DIR" ]] && { echo "dir $DIR already exists"; return 1; }
  mkdir -p torrent/"$IMGNAME" && cp -v --reflink=auto "$TAR" torrent/"$IMGNAME"
  md5sum "$DIR"/*.bz2 > "$DIR"/md5sum
  createtorrent -a udp://tracker.opentrackr.org -p 1337 -c "NextCloudPi. Nextcloud ready to use image" "$DIR" "$DIR".torrent
  transmission-remote -w $(pwd)/torrent -a "$DIR".torrent
}

function generate_changelog()
{
  git log --graph --oneline --decorate \
    --pretty=format:"[%<(13)%D](https://github.com/BstAA/nextcloudpi/commit/%h) (%ad) %s" --date=short | \
    grep 'tag: v' | \
    sed '/HEAD ->\|origin/s|\[.*\(tag: v[0-9]\+\.[0-9]\+\.[0-9]\+\).*\]|[\1]|' | \
    sed 's|* \[tag: |\n[|' > changelog.md
}

function upload_ftp()
{
  local IMGNAME="$1"
  echo -e "\n\e[1m[ Upload FTP ]\e[0m"
  echo "* $IMGNAME..."
  [[ -f torrent/"$IMGNAME"/"$IMGNAME".tar.bz2 ]] || { echo "No image file found, abort"; return 1; }
  [[ "$FTPPASS" == "" ]] && { echo "No FTPPASS variable found, abort"; return 0; }

  cd torrent

  ftp -np ftp.ownyourbits.com <<EOF
user root@ownyourbits.com $FTPPASS
mkdir testing
mkdir testing/$IMGNAME
cd testing/$IMGNAME
binary
rm  $IMGNAME.torrent
put $IMGNAME.torrent
bye
EOF
  cd -
  cd torrent/$IMGNAME

  ftp -np ftp.ownyourbits.com <<EOF
user root@ownyourbits.com $FTPPASS
cd testing/$IMGNAME
binary
rm  $IMGNAME.tar.bz2
put $IMGNAME.tar.bz2
rm  md5sum
put md5sum
bye
EOF
  cd -
}

upload_all()
{
  test -d output || { echo "No uploads found. Nothing to do"; return; }
  [[ "$FTPPASS" == "" ]] && {
    echo -e "\e[1mNo FTPPASS variable found, FTP won't work.\nYou can ^C to cancel now\e[0m"
      sleep 5
    }

  mkdir -p archive
  for img in $(find output -name '*.tar.bz2'); do
    upload_ftp "${img}" && mv "${img}" archive
  done
}


function test_image()
{
  local IMG="$1"
  local IP="$2"
  echo -e "\e[1m\n[ QEMU Test ]\e[0m"
  [[ -f "$IMG" ]] || { echo "No image file found, abort"; return 1; }
  launch_qemu "$IMG" &
  sleep 10
  wait_SSH "$IP"
  sleep 180            # Wait for the services to start. Improve this ( wait HTTP && trusted domains )
  tests/tests.py "$IP"
  ssh_pi "$IP" sudo halt
  wait
}

# License
#
# This script is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This script is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this script; if not, write to the
# Free Software Foundation, Inc., 59 Temple Place, Suite 330,
# Boston, MA  02111-1307  USA


updates/1.16.0.sh
#!/bin/bash

set -e

## BACKWARD FIXES ( for older images )

source /usr/local/etc/library.sh

# all images

# restore smbclient after dist upgrade
apt_install php-smbclient exfat-fuse exfat-utils

# install lsb-release
apt_install lsb-release

# tweak previews
ncc config:app:set preview jpeg_quality --value="60"

# missed some sources
sed -i "s/stretch/$RELEASE/g" /etc/apt/sources.list.d/* &>/dev/null || true

# docker images only
[[ -f /.docker-image ]] && {
:
}

# for non docker images
[[ ! -f /.docker-image ]] && {
  # Update btrfs-sync
  wget -q https://raw.githubusercontent.com/nachoparker/btrfs-sync/master/btrfs-sync -O /usr/local/bin/btrfs-sync
  chmod +x /usr/local/bin/btrfs-sync

  # work around dhcpcd Raspbian bug
  # https://lb.raspberrypi.org/forums/viewtopic.php?t=230779
  # https://github.com/BstAA/nextcloudpi/issues/938
  test -f /usr/bin/raspi-config && {
    apt-get update
    apt_install haveged
    systemctl enable haveged.service
  }

  # Update btrfs-snp
  wget https://raw.githubusercontent.com/nachoparker/btrfs-snp/master/btrfs-snp -O /usr/local/bin/btrfs-snp
  chmod +x /usr/local/bin/btrfs-snp
}

exit 0


updates/1.18.0.sh
#!/bin/bash

set -e

## BACKWARD FIXES ( for older images )

source /usr/local/etc/library.sh # sets NCVER PHPVER RELEASE

# all images

# restore sources in stretch
sed -i "s/buster/$RELEASE/g" /etc/apt/sources.list.d/* &>/dev/null || true

# restore smbclient after dist upgrade
apt_install php${PHPVER}-gmp

# Update modsecurity config file only if user is already in buster and
# modsecurity is used.
# https://github.com/BstAA/nextcloudpi/issues/959
is_active_app modsecurity && run_app modsecurity

# fix armbian disabling unattended-upgrades
is_active_app unattended-upgrades && run_app unattended-upgrades

# groupfolders fix
install_app nc-backup

# docker images only
[[ -f /.docker-image ]] && {
  :
}

# for non docker images
[[ ! -f /.docker-image ]] && {
  # fix fail2ban with UFW
  mkdir -p /etc/systemd/system/fail2ban.service.d/
  cat > /etc/systemd/system/fail2ban.service.d/touch-ufw-log.conf <<EOF
[Service]
ExecStartPre=/bin/touch /var/log/ufw.log
EOF
}

exit 0


ncp-web/index.php
<?php
/*
 NextCloudPi Web Panel frontend

 Copyleft 2017 by Ignacio Nunez Hernanz <nacho _a_t_ ownyourbits _d_o_t_ com>
 GPL licensed (see end of file) * Use at your own risk!

 More at https://ownyourbits.com/2017/02/13/nextcloud-ready-raspberry-pi-image/

**/

// redirect to activation first time
ob_start();
exec("a2query -s ncp-activation", $output, $ret);
if ($ret == 0) {
  header("Location: activate");
  exit();
}

?>

<!DOCTYPE html>
<html class="ng-csp" data-placeholder-focus="false" lang="en">
<head>
    <meta charset="utf-8">
    <title>NextCloudPi Panel</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
    <meta name="referrer" content="never">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=1.0">
    <meta name="mobile-web-app-capable" content="yes">
<?php
    ini_set('session.cookie_httponly', 1);
    if (isset($_SERVER['HTTPS']))
      ini_set('session.cookie_secure', 1);
    session_start();

    include('elements.php');
    $modules_path = '/usr/local/etc/ncp-config.d/';
    $l10nDir = "l10n";

    // security headers
    header("Content-Security-Policy: default-src 'none'; script-src 'self'; connect-src 'self'; img-src 'self'; style-src 'self'; object-src 'self';");
    header("X-XSS-Protection: 1; mode=block");
    header("X-Content-Type-Options: nosniff");
    header("X-Robots-Tag: none");
    header("X-Permitted-Cross-Domain-Policies: none");
    header("X-Frame-Options: DENY");
    header("Cache-Control: no-cache");
    header('Pragma: no-cache');
    header('Expires: -1');

    // HTTP2 push headers
    header("Link: </js/minified.js>; rel=preload; as=script;,</js/ncp.js>; rel=preload; as=script;,</css/ncp.css>; rel=preload; as=style;,</img/ncp-logo.svg>; rel=preload; as=image;, </img/loading-small.gif>; rel=preload; as=image;, rel=preconnect href=ncp-launcher.php;");

  ?>
    <link rel="icon" type="image/png" href="img/favicon.png"/>
    <link rel="stylesheet" href="css/ncp.css">
</head>
<body id="body-user">
<?php
  require("L10N.php");
  try {
    $l = new L10N($_SERVER["HTTP_ACCEPT_LANGUAGE"], $l10nDir, $modules_path);
  } catch (Exception $e) {
    die("<p class='error'>Error while loading localizations!</p>");
  }
?>
<noscript>
    <div id="nojavascript">
        <div>
          <?php sprintf($l->__("This application requires JavaScript for correct operation. Please %s enable JavaScript %s and reload the page."),
              "<a href=\"http://enable-javascript.com/\" target=\"_blank\" rel=\"noreferrer\">", "</a>"); ?>
        </div>
    </div>
</noscript>
<div id="notification-container">
  <?php
    exec("ncp-test-updates", $output, $ret);
    if ($ret == 0) {
      $version  = "v0.0.0";
      $ver_file = '/var/run/.ncp-latest-version';
      if (file_exists($ver_file))
        $version = file_get_contents($ver_file);
      echo '<div id="notification">';
      echo '<div id="update-notification" class="row type-error closeable">';
      echo "version " . $version . " is available";
      echo '<a class="action close icon-close" href="#" alt="Dismiss"></a>';
      echo '</div>';
      echo '</div>';
    }
  ?>
</div>

<?php
  if (file_exists('wizard') && !file_exists('wizard.cfg')) {
    echo <<<HTML
    <div id="first-run-wizard">
      <div class='dialog'>
        <br><br>
        <h2 id="config-box-title">NextCloudPi First Run</h2>
        <p>Click to start the configuration wizard</p>
        <br>
        <a href="wizard"><img class="wizard-btn" src="wizard/img/ncp-logo.svg" class="wizard"></a>
        <br>
        <a href="wizard"> <button type="button" class="wizard-btn" id="go-wizard">{$l->__("run")}  </button></a>
        <button type="button" class="first-run-close" id="skip-wizard" >{$l->__("skip")} </button>
        <button type="button" class="first-run-close close-dialog-x">{$l->__("close")}</button>
        <br><br>
      </div>
    </div>
HTML;
    touch('wizard.cfg');
  }
?>

  <div id="confirmation-dialog" class="hidden">
    <div class='dialog'>
      <br><br>
      <h2 id="config-box-title">Are you sure?</h2>
      <br>
      <p>Click OK to confirm this operation</p>
      <br>
      <button type="button" id="confirmation-dialog-ok"> OK </button>
      <button type="button" class="confirmation-dialog-close"> Cancel </button>
      <button type="button" class="confirmation-dialog-close close-dialog-x">Close</button>
      <br><br>
    </div>
  </div>

  <header role="banner"><div id="header">
    <div id="header-left">
      <a href="https://ownyourbits.com" id="nextcloudpi" target="_blank" tabindex="1">
        <div class="logo-icon">
           <h1 class="hidden-visually">NextCloudPi</h1>
        </div>
      </a>
      <a id=versionlink target="_blank" href="https://github.com/BstAA/nextcloudpi/blob/master/changelog.md">
        <?php
          $version  = "v0.0.0";
          $ver_file = "/usr/local/etc/ncp-version";
          if (file_exists($ver_file))
            $version = file_get_contents($ver_file);
          echo $version;
      ?>
      </a>
      <?php
        // language selection drop
        $selected_lang=$l->load_language_setting();

        $fh = fopen('langs.cfg', 'r');
        echo "<select id=\"language-selection\" name=\"language-selection\">";
        while ($line = fgets($fh)) {
          echo "<option value='" . $line . "' ";
          if( $line == $selected_lang )
            echo "selected='selected'";
          echo ">". $line ."</option>";
        }
        echo "<option value=\"[new]\">new..</option>";
        echo "</select>";
        fclose($fh);
    ?>
    </div>
    <div id="header-right">
      <div id="search-btn" title="<?php echo $l->__("Search"); ?>">
          <div class="expand search-div">
              <div class="icon-search"></div>
              &nbsp; &nbsp;
              <input type="text" id="search-box" placeholder="find ncp-app" size="40">
          </div>
      </div>
      <a href="https://ownyourbits.com" id="nextcloud-btn" target="_blank" tabindex="1" title="<?php echo $l->__("Launch Nextcloud"); ?>">
        <div id="nc-button">
            <div class="expand">
                <div class="icon-nc-white"></div>
            </div>
        </div>
      </a>
      <div id="dashboard-btn" title="<?php echo $l->__("View System Info"); ?>">
          <div class="expand">
              <div class="icon-dashboard"></div>
          </div>
      </div>
      <div id="backups-btn" title="<?php echo $l->__("Backups and snapshots"); ?>">
          <div class="expand">
              <div class="icon-backups"></div>
          </div>
      </div>
      <div id="config-btn" title="<?php echo $l->__("Nextcloud Configuration"); ?>">
          <div class="expand">
              <div class="icon-config"></div>
          </div>
      </div>
      <a href="wizard" title="<?php echo $l->__("NextCloudPi Wizard"); ?>">
        <div class="wizard-btn">
          <div class="expand">
            <div class="icon-wizard-white"></div>
          </div>
        </div>
      </a>
      <a href="https://github.com/BstAA/nextcloudpi/wiki" target="_blank" tabindex="1"  title="<?php echo $l->__("NextCloudPi Wiki"); ?>">
        <div id="nc-button">
            <div class="expand">
                <div class="icon-nc-info"></div>
            </div>
        </div>
      </a>
      <div id="poweroff" title="<?php echo $l->__("Power Off"); ?>">
          <div class="expand">
              <div class="icon-power-white"></div>
          </div>
      </div>
  </div>
</header>

  <div id="content-wrapper">
	<div id="content" class="app-files" role="main">
        <div id='overlay' class="hidden"></div>
		<div id="app-navigation">
        	<div id="ncp-options">
              <?php echo print_sidebar($l, false); ?>
            </div>
          </div>

      <div id="app-content">
        <div id="app-navigation-toggle" class="icon-menu hidden"></div>

        <div id="config-wrapper">
          <?php echo print_config_forms($l); ?>
        </div>

        <div id="dashboard-wrapper" class="content-box <?php if(array_key_exists('app',$_GET) && ($_GET['app'] != 'dashboard')) echo 'hidden';?>">
          <h2 class="text-title"><?php echo $l->__("System Info"); ?></h2>
          <div id="dashboard-suggestions" class="table-wrapper"></div>
          <div id="dashboard-table" class="outputbox table-wrapper"></div>
          <div id="loading-info-gif" class="loading-section-gif"> <img src="img/loading-small.gif"> </div>
        </div>

        <div id="backups-wrapper" class="content-box <?php if($_GET['app'] != 'backups') echo 'hidden';?>">
          <h2 class="text-title"><?php echo $l->__("Backups"); ?></h2>
          <div id="backups-content" class="table-wrapper"></div>
          <div id="loading-backups-gif" class="loading-section-gif"> <img src="img/loading-small.gif"> </div>
        </div>

        <div id="nc-config-wrapper" class="content-box <?php if($_GET['app'] != 'config') echo 'hidden';?>">
          <h2 class="text-title"><?php echo $l->__("Nextcloud configuration"); ?></h2>
          <div id="nc-config-box" class="table-wrapper">
<?php
  // print Nextcloud config table
  include( '/var/www/nextcloud/config/config.php' );

  echo '<table class="dashtable">';
  foreach ( $CONFIG as $key => $val )
  {
    echo '<tr>';
    if ( gettype( $val ) == "array" )
    {
      echo "<td>$key</td>";
      echo "<td><table>";
      foreach ( $val as $k => $v )
        echo "<tr><td>$k</td><td class=\"val-field\">$v</td></tr>";
      echo '</table></td>';
    }
    else
      echo "<td>$key</td><td class=\"val-field\">$val</td>";
    echo '</tr>';
  }
  echo '</table>';
?>
        </div>
        </div>
    </div>

  <div id="poweroff-dialog" class='dialog primary hidden'>
      <div id='poweroff-option_shutdown' class='button big-button'>
         <img class="wizard-btn" src="img/poweroff.svg">&nbsp;&nbsp;shut down
      </div>
      <div id='poweroff-option_reboot'   class='button big-button'>
         <img class="wizard-btn" src="img/reboot.svg">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;reboot&nbsp;&nbsp;&nbsp;
      </div>
  </div>

  <?php
    include('csrf.php');
    echo '<input type="hidden" id="csrf-token"     name="csrf-token"     value="' . getCSRFToken() . '"/>';
    echo '<input type="hidden" id="csrf-token-ui"  name="csrf-token-ui"  value="' . getCSRFToken() . '"/>';
    echo '<input type="hidden" id="csrf-token-cfg" name="csrf-token-cfg" value="' . getCSRFToken() . '"/>';
  ?>
    <script src="js/minified.js"></script>
    <script src="js/ncp.js"></script>
</body>
</html>

<!--
 License

 This script is free software; you can redistribute it and/or modify it
 under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This script is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this script; if not, write to the
 Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 Boston, MA  02111-1307  USA
-->
